### ISSUE TITLE
Saving vector of dictionaries initialized as #undef gives "This should not have happened"

### ISSUE THREAD

**Issue by Dale-Muccignat** [JuliaIO/JLD2.jl#510]
Before we start, I'd like to state that I know this is very specific but I ran into this while saving heaps of data for my work. I often initialize variables with undef before filling them out. Sometimes they don't need to be populated but it was still saving the variable.

Essentially this is what happens as a mwe:
```
julia> x=transportlist_train = Vector{Dict{String,Float64}}(undef, 10)
10-element Vector{Dict{String, Float64}}:
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef
 #undef

julia> @save "test.jld2" x

julia> @load "test.jld2"
ERROR: This should not have happened
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] read_obj_start
    @ ~/.julia/packages/JLD2/u57Vt/src/object_headers.jl:104 [inlined]
  [3] isgroup(f::JLD2.JLDFile{JLD2.MmapIO}, roffset::JLD2.RelOffset)
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/object_headers.jl:127
  [4] load_dataset(f::JLD2.JLDFile{JLD2.MmapIO}, offset::JLD2.RelOffset)
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/datasets.jl:16
  [5] jlconvert
    @ ~/.julia/packages/JLD2/u57Vt/src/data/writing_datatypes.jl:314 [inlined]
  [6] jlconvert(#unused#::JLD2.ReadRepresentation{Dict{String, Float64}, JLD2.CustomSerialization{Array, JLD2.RelOffset}}, f::JLD2.JLDFile{JLD2.MmapIO}, ptr::Ptr{Nothing}, header_offset::JLD2.RelOffset)
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/data/custom_serialization.jl:56
  [7] macro expansion
    @ ~/.julia/packages/JLD2/u57Vt/src/dataio.jl:70 [inlined]
  [8] macro expansion
    @ ./simdloop.jl:77 [inlined]
  [9] read_array!
    @ ~/.julia/packages/JLD2/u57Vt/src/dataio.jl:68 [inlined]
 [10] read_array(f::JLD2.JLDFile{JLD2.MmapIO}, dataspace::JLD2.ReadDataspace, rr::JLD2.ReadRepresentation{Dict{String, Float64}, JLD2.CustomSerialization{Array, JLD2.RelOffset}}, layout::JLD2.DataLayout, filters::JLD2.FilterPipeline, header_offset::JLD2.RelOffset, attributes::Vector{JLD2.ReadAttribute})
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/datasets.jl:410
 [11] read_data(f::JLD2.JLDFile{JLD2.MmapIO}, rr::Any, read_dataspace::Tuple{JLD2.ReadDataspace, JLD2.RelOffset, JLD2.DataLayout, JLD2.FilterPipeline}, attributes::Vector{JLD2.ReadAttribute})
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/datasets.jl:240
 [12] read_data(f::JLD2.JLDFile{JLD2.MmapIO}, dataspace::JLD2.ReadDataspace, datatype_class::UInt8, datatype_offset::Int64, layout::JLD2.DataLayout, filters::JLD2.FilterPipeline, header_offset::JLD2.RelOffset, attributes::Vector{JLD2.ReadAttribute})
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/datasets.jl:194
 [13] load_dataset(f::JLD2.JLDFile{JLD2.MmapIO}, offset::JLD2.RelOffset)
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/datasets.jl:125
 [14] getindex(g::JLD2.Group{JLD2.JLDFile{JLD2.MmapIO}}, name::String)
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/groups.jl:109
 [15] read
    @ ~/.julia/packages/JLD2/u57Vt/src/JLD2.jl:479 [inlined]
 [16] (::var"#223#224")(f::JLD2.JLDFile{JLD2.MmapIO})
    @ Main ~/.julia/packages/JLD2/u57Vt/src/loadsave.jl:146
 [17] jldopen(f::Function, args::String; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/loadsave.jl:4
 [18] jldopen(f::Function, args::String)
    @ JLD2 ~/.julia/packages/JLD2/u57Vt/src/loadsave.jl:1
 [19] top-level scope
    @ ~/.julia/packages/JLD2/u57Vt/src/loadsave.jl:145
 [20] top-level scope
    @ ~/.julia/packages/CUDA/YIj5X/src/initialization.jl:208
```

I have since seen that I should be doing something like this instead
```
julia> [typeof(data)() for _ in 1:10]
10-element Vector{Dict{String, Dict{Float64, Dict{Float64, Dict{Float64, Dict{String, Vector{Vector{Float64}}}}}}}}:
 Dict()
 Dict()
 Dict()
 Dict()
 Dict()
 Dict()
 Dict()
 Dict()
 Dict()
 Dict()
```

So happy to leave this to user error but it is something I ran into that was hard to track down so worth documenting.

**Comment by JonasIsensee** [comment]
Thanks for the report.
I'm glad you found a way to avoid the bug already.

**Comment by JonasIsensee** [comment]
It is fixed on v0.4.40
